src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,674 @@
+            if all_types:
+                chosen_attributes = all_types
+                cluster = Clusters.ClusterObjects.ALL_CLUSTERS[chosen_attributes[0].cluster_id]
+                break
+        else:
+            print(f"Attribute not found on device: {desired_attribute_type}")
+            cluster = None
+
+        endpoint = None
+        for endpoint in self.endpoints:
+            for cluster, clusterdata in self.endpoints[endpoint].items():
+                all_types = await self.all_type_attributes_for_cluster(cluster, desired_attribute_type)
+                attributes_of_type = set(all_types)
+                attributes_of_type_on_device = attributes_of_type.intersection(set(clusterdata.keys()))
+                if attributes_of_type_on_device:
+                    return True
+        return False
+
+    async def read_unsupported_attribute(self):
+        found_unsupported = False
+        for endpoint_id, endpoint in self.endpoints.items():
+            print(endpoint_id, next(iter(endpoint)))
+            if found_unsupported:
+                break
+            for cluster_type, cluster in endpoint.items():
+                if global_attribute_ids.cluster_id_type(cluster_type.id) != global_attribute_ids.ClusterIdType.kStandard:
+                    continue
+
+                all_attrs = set(list(ClusterObjects.ALL_ATTRIBUTES[cluster_type.id].keys()))
+                dut_attrs = set(cluster[cluster_type.Attributes.AttributeList])
+
+                unsupported = [id for id in list(all_attrs - dut_attrs) if global_attribute_ids.attribute_id_type(id)
+                               == global_attribute_ids.AttributeIdType.kStandardNonGlobal]
+                if unsupported:
+                    result = await self.read_single_attribute_expect_error(
+                        endpoint=endpoint_id,
+                        cluster=cluster_type,
+                        attribute=ClusterObjects.ALL_ATTRIBUTES[cluster_type.id][unsupported[0]], error=Status.UnsupportedAttribute)
+                    asserts.assert_true(isinstance(result.Reason, InteractionModelError),
+                                        msg="Unexpected success reading invalid attribute")
+                    found_unsupported = True
+                    return
+
+    @async_test_body
+    async def test_TC_IDM_2_2(self):
+        # Test Setup
+        await self.setup_class_helper(allow_pase=False)
+
+        all_clusters = [cluster for cluster in Clusters.ClusterObjects.ALL_ATTRIBUTES]
+
+        self.device_clusters = self.all_device_clusters()
+        self.device_attributes = self.all_device_attributes()
+        self.all_supported_clusters = [cluster for cluster in Clusters.__dict__.values(
+        ) if inspect.isclass(cluster) and issubclass(cluster, ClusterObjects.Cluster)]
+
+        self.print_step(0, "Commissioning - already done")
+
+        wildcard_descriptor = await self.default_controller.ReadAttribute(self.dut_node_id, [(Clusters.Descriptor)])
```

Comment:
I don't think wildcard_descriptor and endpoints get used anywhere. Rest of the code seems to use self.endpoints, which is the full wildcard.

Reply:
Removed